File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:

        l,r = 0, len(nums)-1 #l=0, r=5
        
        
        while l <= r:

            m = (l+r) // 2 #m=2
            if target > nums[m]: #target index 345
                l = m + 1
            elif target < nums[m]: #target index 1
                r = m - 1
            else:
                return m
        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int ROW = matrix.size();
        int COL = matrix[0].size();

        int top = 0, bot = ROW - 1;
        while (top<=bot) {
            int row = top + (bot-top)/2;
            if (target > matrix[row][COL-1]) top = row+1;
            else if (target < matrix[row][0]) bot = row-1;
            else break; 
        }
        if (top>bot) return false;
        int l = 0, r = COL-1;
        int row = top + (bot-top)/2;

        while (l<=r) {
            int m = l + (r-l)/2;
            if (target > matrix[row][m]) l=m+1;
            else if (target < matrix[row][m]) r=m-1;
            else return true;
        }
        return false;
    }
};
